views/default/input/form.php

```diff
@@ -23,17 +25,40 @@
 $vars = array_merge($defaults, $vars);
 
 $vars['class'] = elgg_extract_class($vars, 'elgg-form');
-$vars['action'] = elgg_normalize_url($vars['action']);
-$vars['method'] = strtolower($vars['method']);
 
-$body = $vars['body'];
+$vars['action'] = elgg_normalize_url(elgg_extract('action', $vars, ''));
+unset($vars['action_name']);
+
+$vars['method'] = strtolower(elgg_extract('method', $vars, 'POST'));
+
+$body = elgg_extract('body', $vars, '');
 unset($vars['body']);
 
-// Generate a security header
-if (!$vars['disable_security']) {
-	$body = elgg_view('input/securitytoken') . $body;
+// Add CSRF tokens
+$tokens = '';
+if (!elgg_extract('disable_security', $vars, false)) {
+	$tokens = elgg_view_input('securitytoken');
 }
 unset($vars['disable_security']);
-unset($vars['action_name']);
 
-echo elgg_format_element('form', $vars, "<fieldset>$body</fieldset>");
+$header = '';
+if (isset($vars['header'])) {
+	$header = elgg_format_element('header', [
+		'class' => 'elgg-form-header',
+	], $vars['header']);
+	unset($vars['header']);
+}
+
+$body = elgg_format_element('fieldset', [
+	'class' => 'elgg-form-body',
+], $tokens . $body);
+
+$footer = '';
+if (isset($vars['footer'])) {
```

Comment:
I initially thought I would make the PR against 2.x, but it's not going to work. This is where the deferred footer should go.